plugins/GSdx/GSTextureCache.cpp

```diff
@@ -2043,3 +2034,164 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+}
+
+uint32* GSTextureCache::Palette::GetClut() {
+	return m_clut;
+}
+
+GSTexture* GSTextureCache::Palette::GetPaletteGSTexture() {
+	return m_tex_palette;
+}
+
+// GSTextureCache::PaletteMap
+
+// Default constructor, reverses space in the multimaps
+GSTextureCache::PaletteMap::PaletteMap() {
+	for (auto& multimap : m_multimaps) {
+		multimap.reserve(MAX_SIZE);
+	}
+}
+
+// Hashes the content of the clut
+size_t GSTextureCache::PaletteMap::HashClut(uint16 pal, const uint32* clut) {
+	size_t clut_hash = 3831179159;
+	for (uint16 i = 0; i < pal; i += 16) {
+		clut_hash = (clut_hash + 1488000301) ^ (clut[i] + 33644011);
+		clut_hash = (clut_hash + 3831179159) ^ (clut[i + 1] + 47627467);
+		clut_hash = (clut_hash + 3659574209) ^ (clut[i + 2] + 577038523);
+		clut_hash = (clut_hash + 33644011) ^ (clut[i + 3] + 3491555267);
+
+		clut_hash = (clut_hash + 777771959) ^ (clut[i + 4] + 3301075993);
+		clut_hash = (clut_hash + 4019618579) ^ (clut[i + 5] + 4186992613);
+		clut_hash = (clut_hash + 3465668953) ^ (clut[i + 6] + 3043435883);
+		clut_hash = (clut_hash + 3494478943) ^ (clut[i + 7] + 3441897883);
+
+		clut_hash = (clut_hash + 3432010979) ^ (clut[i + 8] + 2167922789);
+		clut_hash = (clut_hash + 1570862863) ^ (clut[i + 9] + 3401920591);
+		clut_hash = (clut_hash + 1002648679) ^ (clut[i + 10] + 1293530519);
+		clut_hash = (clut_hash + 551381741) ^ (clut[i + 11] + 2539834039);
+
+		clut_hash = (clut_hash + 3768974459) ^ (clut[i + 12] + 169943507);
+		clut_hash = (clut_hash + 862380703) ^ (clut[i + 13] + 2906932549);
+		clut_hash = (clut_hash + 3433082137) ^ (clut[i + 14] + 4234384109);
+		clut_hash = (clut_hash + 2679083843) ^ (clut[i + 15] + 2719605247);
+	}
+	return clut_hash;
+}
+
+// Retrieves the palette with the desired clut
+std::shared_ptr<GSTextureCache::Palette> GSTextureCache::PaletteMap::LookupPalette(uint16 pal) {
+	// Choose which hash map search into:
+	//    pal == 16  : index 0
+	//    pal == 256 : index 1
+	auto& multimap = m_multimaps[pal == 16 ? 0 : 1];
+
+	const uint32* clut = (const uint32*)m_renderer->m_mem.m_clut;
+
+	// Search for the clut hash into the hash map
+	size_t clut_hash = HashClut(pal, clut);
+
+	// Retrieve all the relevant Palette(s)
+	auto it1 = multimap.equal_range(clut_hash);
+	uint16 palette_size = pal * sizeof(uint32);
+	for (auto it2 = it1.first; it2 != it1.second; ++it2) {
+		// Hash check passed, compare Palette clut content
+		std::shared_ptr<Palette> palette = it2->second;
+		if (GSVector4i::compare64(palette->GetClut(), clut, palette_size)) {
+			// Clut content match, HIT
+			return palette;
+		}
+		// Same clut hash but content mismatch, FALSE HIT
+		GL_PERF("TC, %u-bit PaletteMap (Size %u): FALSE HIT for hash value %u.", palette_size, multimap.size(), clut_hash);
+	}
+
+	// No Palette with matching clut content hash, MISS
+
+	if (multimap.size() > MAX_SIZE) {
+		// If the multimap is too big, try to clean it by disposing and removing unused palettes, before adding the new one
```

Comment:
If I remove the cleaning logic performed if the size of the map exceeds a given threshold, the size of the map would only increase with time, as the palettes remains in memory independently of the lifecycle of its users (the `Source` objects).
The expected result of this cleaning logic is that many palettes in the map gets cleared in a single shot, because the number of `Source` objects that can be active on screen at the same time is much smaller than `MAX_SIZE` leaving many palettes unused at each frame (but can be reused at the next, maybe), and thus at the next lookup call, if the result is miss + insert new palette, the cleaning logic is not triggered as the map was reduced greatly in size.
In the tests I performed, the logic was rarely hit, but the maximum size could be tuned to have a more aggressive cleaning of the unused palettes, in case the lookup slows down too much the rendering (which I didn't experience, though); or vice versa.